File: Graphs/JS/weightedGraph.js
const Node = function (val, weight) {
  this.node = val;
  this.weight = weight;
};

const BHTPQ = class {
  constructor() {
    this.pq = [];
  }

  enqueue(val, weight) {

    if (val && weight >= 0) {
      const node = new Node(val, weight);
      if (this.pq[0]) {
        this.pq[this.pq.length] = node;
        this.heapUp();
      } else this.pq[0] = node;
    }

  }

  heapUp() {

    let cIdx = this.pq.length - 1,
      pIdx = Math.floor((cIdx - 1) / 2),
      tmp;

    while (this.pq[pIdx] && this.pq[pIdx].weight > this.pq[cIdx].weight) {
      tmp = this.pq[pIdx];
      this.pq[pIdx] = this.pq[cIdx];
      this.pq[cIdx] = tmp;

      cIdx = pIdx;
      pIdx = Math.floor((cIdx - 1) / 2);
    }

  }

  dequeue() {
    if (this.pq) {
      const min = this.pq[0];
      this.pq[0] = this.pq.pop();
      this.heapDown();
      return min;
    }
  }

  heapDown() {
    let pIdx = 0,
      leftCIdx = pIdx * 2 + 1,
      rightCIdx = leftCIdx + 1,
      smallCIdx,
      tmp;

    while (
      (this.pq[leftCIdx] && this.pq[leftCIdx].weight < this.pq[pIdx].weight) ||
      (this.pq[rightCIdx] && this.pq[rightCIdx].weight < this.pq[pIdx].weight)
    ) {
      if (!rightCIdx) smallCIdx = leftCIdx;
      else
        smallCIdx =
          this.pq[leftCIdx].weight < this.pq[rightCIdx].weight
            ? leftCIdx
            : rightCIdx;

      tmp = this.pq[pIdx];
      this.pq[pIdx] = this.pq[smallCIdx];
      this.pq[smallCIdx] = tmp;

      pIdx = smallCIdx;
      leftCIdx = pIdx * 2 + 1;
      rightCIdx = leftCIdx + 1;
    }
  }
};

const WUGraph = class {
  constructor() {
    this.adjacencyList = {};
  }

  addVertex(vertex) {
    if (!this.adjacencyList[vertex]) this.adjacencyList[vertex] = [];
  }

  addEdge(vertex1, vertex2, weight) {
    if (this.adjacencyList[vertex1] && this.adjacencyList[vertex2] && weight) {
      if (!this.adjacencyList[vertex1].includes(vertex2)) {
        this.adjacencyList[vertex1][this.adjacencyList[vertex1].length] =
          new Node(vertex2, weight);
      }
      if (!this.adjacencyList[vertex1].includes(vertex1)) {
        this.adjacencyList[vertex2][this.adjacencyList[vertex2].length] =
          new Node(vertex1, weight);
      }
    }
  }

  removeVertex(vertex) {
    if (this.adjacencyList[vertex]) {
      this.adjacencyList[vertex].forEach((v) => {
        this.adjacencyList[v.node] = this.adjacencyList[v.node].filter(
          (e) => e.node !== vertex
        );
      });
      delete this.adjacencyList[vertex];
    }
  }

  shortestPath(start, end) {
    const distances = {},
      previous = {},
      queue = new BHTPQ();
    for (const key in this.adjacencyList) {
      key !== start ? (distances[key] = Infinity) : (distances[key] = 0);
      key !== start ? queue.enqueue(key, Infinity) : queue.enqueue(key, 0);
      previous[key] = null;
    }

    let current, pathCost;
    while (queue.pq[0]) {
      current = queue.dequeue().node;

      if (current === end) break;
      if (distances[current] !== Infinity) {
        this.adjacencyList[current].forEach((v) => {
          pathCost = distances[current] + v.weight;
          if (pathCost < distances[v.node]) {
            distances[v.node] = pathCost;
            previous[v.node] = current;
            queue.enqueue(v.node, distances[v.node]);
          }
        });
      }
    }

    const path = [end];
    let tmp = end;
    while (true) {
      if (tmp === start) break;
      tmp = previous[tmp];
      path[path.length] = tmp;
    }
    console.log('path: ', path.reverse());
    return distances[end];
  }
};

const wUG = new WUGraph();

wUG.addVertex('A');
wUG.addVertex('B');
wUG.addVertex('C');
wUG.addVertex('D');
wUG.addVertex('E');
wUG.addVertex('F');

console.log("wUG's adjacencyList: ", wUG.adjacencyList);
wUG.addEdge('A', 'B', 4);
wUG.addEdge('A', 'C', 2);
wUG.addEdge('B', 'E', 3);
wUG.addEdge('C', 'D', 2);
wUG.addEdge('C', 'F', 4);
wUG.addEdge('D', 'E', 3);
wUG.addEdge('D', 'F', 1);
wUG.addEdge('E', 'F', 1);
// console.log("wUG's adjacencyList: ", wUG.adjacencyList);
// wUG.removeVertex('A');
// console.log("wUG's adjacencyList: ", wUG.adjacencyList);

console.log('shortest path method: ', wUG.shortestPath('A', 'E'));

/*
console.log('v: ', v);
 console.log('queue: ', queue);
 console.log('current: ', current);v
    console.log('this.adjacencyList: ', this.adjacencyList)
    console.log('\n distances: ', distances);
    console.log('\n previous: ', previous)
*/


/*
  ///////////////////////////////////////// NODE  /////////////////////////////////////////////////
  1.  CREATE A FUNCTION NAMED Node THAT WILL CREATE NODES FOR BINARY HEAP TREE PRIORITY QUEUE AND WEIGHTED UNDIRECTED GRAPH

      A.  IT TAKES A VALUE (STRING) AND WEIGHT (NUMBER) AS ITS PARAMETERS
      B.  WHEN CALLED WITH THE NEW KEY WORD, IT WILL RETURN A CLASS WITH THE ARGUMENTS PASSED IN AS ITS PROPERTIES

  //////////////////////////////////////// PRIORITY QUEUE  ///////////////////////////////////

  1.  CREATE A BINARY-HEAP-TREE-PRIORITY-QUEUE CLASS NAMED HBTPQ

      A.  IT WILL HAVE A ARRAY AS ITS SOLE PROPERTY NAMED pq

  2.  CREATE A METHOD TO ENQUEUE A NEW NODE INTO pq NAMED enqueue

      A.  IT WILL TAKE A VALUE (STRING) AND WEIGHT (NUMBER) AS ITS PARAMETERS
      B.  IT WILL CREATE A NEW NODE BY CALLING Node WITH THE PARAMATERS PASSED INTO THE METHOD
      C.  IF THERE IS A VALUE IN THE pq ARRAY
          I.    PUSH THE JUST CREATED NODE ONTO THE pq array
          II.   CALL THE heapUp method
      D.  ELSE IF THE pq ARRAY IS EMPTY, PUSH THE JUST CREATED NODE ONTO THE pq array

  3.  CREATE A METHOD TO SORT THE BINARY HEAP TREE PRIORITY QUEUE WHEN A NODE HAS BEEN ADDED TO THE END NAMED heapUp

      A.  CREATE THE cIdx (CHILD INDEX) VARIABLE AND ASSIGN THE VALUE OF THE LAST NODE IN pq
      B.  CREATE THE pIdx (PARENT INDEX) VARIABLE AND ASSIGN THE VALUE OF USING THE FORMULA  (n - 1) / 2  WHEE n IS THE CHILD INDEX; THAT IS cIdx
      C.  LOOP WHILE pIdx IS NOT NULL AND WHILE THE NODE IN THE pq ARRAY AT PARENT INDEX HAS A WEIGHT PROPERTY VALUE THAT IS GREATER
          THAN THE WEIGHT PROPERTY OF THE NODE IN THE pq ARRAY THAT IS AT THE INDEX OF cIdx
          I.    SWAP THE PARENT NODE WITH THE CHILD NODE IN THE pq ARRAY
          II.   ASSIGN THE VALUE OF cIdx TO pIdx
          III.  ASSIGN THE VALUE OF USING THE FORMULA  (n - 1) / 2  WHEE n IS THE CHILD INDEX TO pIdx

  4.  CREATE A METHOD TO DEQUEUE A NODE FROM pq NAMED dequeue

      A.  IF THERE IS A VALUE IN THE pq ARRAY
          I.    CREATE A VARIABLE NAMED min AND ASSIGN THE VALUE OF THE FIRST NODE IN THE pq ARRAY TO IT
          II.   MAKE THE VALUE OF THE FIRST NODE/ELEMENT IN THE pq ARRAY TO BE THE VALUE OF THE LAST NODE/ELEMENT IN THE pq ARRAY
          III.  CALL THE heapDown METHOD
          IV.   RETURN min

  5.  CREATE A METHOD TO SORT THE BINARY HEAP TREE PRIORITY QUEUE WHEN A NODE HAS BEEN REMOVED FROM THE END
      AND ASSIGNED TO THE BEGINNING OF THE BINARY HEAP TREE PRIORITY QUEUE

      A.

  //////////////////////////////////////// GRAPH  ///////////////////////////////////////////

  1. CREATE A WEIGHTED UNDIRECTED GRAPH CALLED WUGraph
  2. CREATE A METHOD TO ADD VERTICES TO THE ADJACENCY LIST OF WUGraph CALLED addVertex
      A. IT TAKES A VALUE (STRING) AND WEIGHT (NUMBER) AS ITS PARAMETERS AND CREATES A VERTEX BY CALLING Node, IF IT DOESN'T ALREADY EXIST
      B. IT ADDS THE VERTEX TO THE ADJACENCY LIST WITH PARAMETER'S VALUE AS THE KEY AND AN ARRAY AS THE KEY'S VALUE

      TIME COMPLEXITY: O(N)
      SPACE COMPLEXITY: O(1)

  3. CREATE A METHOD THAT CREATES AN EDGE BETWEEN TWO VERTICES
      A.  IT TAKES TWO VERTICES AS ITS PARAMETERS
      B.  IF THE TWO VERTICES ARE FOUND IN THE INSTANCE OF THE UNORDERED GRAPH
          I.    IF THE FIRST VERTEX DOES NOT HAVE THE SECOND VERTEX IN ITS ARRAY OF VERTICES
                a.  PUSH THE SECOND VERTEX ONTO THE ARRAY THAT BELONGS TO THE FIRST VERTEX
          II.   IF THE SECOND VERTEX DOES NOT HAVE THE FIRST VERTEX IN ITS ARRAY OF VERTICES
                a.  PUSH THE FIRST VERTEX ONTO THE ARRAY THAT BELONGS TO THE SECOND VERTEX

  4.  CREATE A METHOD TO REMOVE A VERTEX
      A.  IT TAKES A VERTEX AS A PARAMETER
      B.  IF THAT VERTEX IS IN THE ADJACENCY LIST OF THE INSTANCE OF THE WEIGHTED UNORDERED GRAPH
          I.    LOOP OVER THE VERTEX'S ARRAY OF VERTICES
          II.   FOR EACH OF THOSE VERTICES FILTER OUT OF ITS ARRAY THE VERTEX PASSED INTO THE METHOD
          III.  FINALLY, DELETE VERTEX PASSED INTO METHOD FROM THE ADJACENCYLIST
*/

File: Graphs/python/weighted_graph.py
"""
    STEP 7
    Queue
    1.  create a Queue class that will serve as a priority queue, keeping a list of dictionaries
        in sorted order by their priority property

"""


class Queue:
    def __init__(self):
        self.values = []

    def enqueue(self, val, pri):
        if type(val) == str and pri >= 0:
            self.values.append({"val": val, "priority": pri})
            self.values.sort(key=lambda info: info["priority"])

    def dequeue(self):
        node = None
        if self.values:
            node = self.values.pop(0)
        return node


"""
    STEP 1
    Node
    1.  Create a Node class that will provide nodes/vertices for the Weighted_Graph class
        I.  It will take a values and make them properties of the class instance
"""


class Node:
    def __init__(self, val):
        self.val = val


"""
    STEP 2
    Weighted_Graph
    1.  Create a Weighted_Graph class that have and adjacency list property that
        will be a dictionary of all vertices in graph as keys, and each key will have
        a value of that vertex's neighbors as a list of dictionaries with information
        about the neighbor
        2.  add_vertex method will add a vetex to the adjacency list
        3.  create_edge method will create and edge between two vertices
        4.  remove_edge method will remove an edge between two vertices
        5.  remove_vertex method will delete a vertex from the adjacency list
"""


class Weighted_Graph:
    def __init__(self):
        self.adjacency_list = {}

    """
        STEP 3
        NOTE: IN REAL WORLD SCENARIOS OBJECTS/NODES ARE MOST LIKELY USED TO REPRESENT SOME ENTITY
        add_vertex
        1.  if valid argument is passed in
            2.  if a vertex by such a title does not already exist in the adjacency list
                3.  create a vertex/node by calling Node class with the argument/s passed in
                4.  add that vertex/node as a value with the key of the title passed in
        time complexity: O(1)
    """

    def add_vertex(self, vertex_name):
        if type(vertex_name) == str:
            if vertex_name not in self.adjacency_list:
                self.adjacency_list[vertex_name] = {
                    "node": Node(vertex_name),
                    "edges": [],
                }

    """
        STEP 4
        NOTE: the list of edges should be as light as can be for speed and memory efficiency
        create_edge
        1.  if valid arguments are passed in
            2.  if the titles of both vertices exist in adjacency list
                3.  add each other to each others respective adjacency list
        time complexity: O(1)
    """

    def create_edge(self, vertex1_name, vertex2_name, weight):
        if type(vertex1_name) == str and type(vertex2_name) == str:
            if (
                vertex1_name in self.adjacency_list
                and vertex2_name in self.adjacency_list
            ):
                self.adjacency_list[vertex1_name]["edges"].append(
                    {"node": vertex2_name, "weight": weight}
                )
                self.adjacency_list[vertex2_name]["edges"].append(
                    {"node": vertex1_name, "weight": weight}
                )

    """
        STEP 5
        remove_edge
        1.  if valid arguments are passed in
            2.  if the titles of both vertices exist in adjacency list
                3.  remove each other from each others respective adjacecy list
                    by filtering out the edge to remove
        time complexity: O(n)
    """

    def remove_edge(self, vertex1_name, vertex2_name):
        if type(vertex1_name) == str and type(vertex2_name) == str:
            if (
                vertex1_name in self.adjacency_list
                and vertex2_name in self.adjacency_list
            ):
                vertex1_updated_edges = []
                for _, edge in enumerate(self.adjacency_list[vertex1_name]["edges"]):
                    if edge["node"] != vertex2_name:
                        vertex1_updated_edges.append(edge)
                self.adjacency_list[vertex1_name]["edges"] = vertex1_updated_edges
                vertex2_updated_edges = []
                for _, edge in enumerate(self.adjacency_list[vertex2_name]["edges"]):
                    if edge["node"] != vertex1_name:
                        vertex2_updated_edges.append(edge)
                self.adjacency_list[vertex2_name]["edges"] = vertex2_updated_edges

    """
        STEP 6
        1.  if valid argument is passed in
        2.  if the title of the vertex to be deleted exist in the adjacecy list
            3.  iterate over the list of the edges for the vertex that is to be deleted
                4.  for each vertex in list, iterate over that vertex's list of edges
                    5.  remove the node to be deleted from the list
            6.  delete from the adjacency list the vertex passed in as an argument
    """

    def remove_vertex(self, vertex_name):
        if type(vertex_name) == str:
            if vertex_name in self.adjacency_list:
                for _, edge in enumerate(self.adjacency_list[vertex_name]["edges"]):
                    updated_edges = []
                    edges = self.adjacency_list[edge["node"]]["edges"]
                    for _, other_edge in enumerate(edges):
                        if other_edge["node"] != vertex_name:
                            updated_edges.append(other_edge)
                    self.adjacency_list[edge["node"]]["edges"] = updated_edges
                del self.adjacency_list[vertex_name]

    """
        STEP 8
        NOTE:   shortest parth from one vertex to another in graph
        shortest_path()
        INITIATE DATA STRUCTURES WITH DATA
            1.  create dictionary, call it distances. While iterating over the adjacecy list, it
                will create a key in dictionary, representing each vertex in adjacency list.
                Moreover, it will initiate each key with a value of Infinity, except for the
                starting vertex, which will have a value of zero.
            2.  create a priority queue, call it queue. While iterating over adjacency list, it
                will place the vertex with a priority of Infinity in the priority queue, except
                for the starting vertex, which will have a priority of zero.
            3.  create another dictionary, call it previous. While iterating over the adjacency list,
                it will create a key in dictionary, representing each vertex in the adjacency list.
                Moreover, it will it will initiate each key will a value of None.

        LOOP OVER PRIORITY QUEUE WHILE QUEUE IS NOT EMPTY OR DESTINATION VERTEX IS FOUND
            4.  loop while queue has an item in it
            5.  remove the first item
            6.  check if it is the destination vertex and return if it is
            7.  get the payload from the adjacency for the popped item
            8.

    """

    def shortest_path(self, start_vertex_name, end_vertex_name):
        #   initiate data structures with data
        queue, distances, previous = Queue(), {}, {}
        for _, vertex_name in enumerate(self.adjacency_list):
            if vertex_name == start_vertex_name:
                queue.enqueue(start_vertex_name, 0)
                distances[start_vertex_name] = 0
            else:
                queue.enqueue(vertex_name, float("inf"))
                distances[vertex_name] = float("inf")
            previous[vertex_name] = None
        print("queue: ", queue.values)
        print("distances: ", distances)
        print("previous: ", previous)
        #  loop while priority queue has something in the queue
        while queue:
            #  remove the first item in queue
            node = queue.dequeue()
            #   if destination vertex is dequeued from priority queue
            if node["val"] == end_vertex_name:
                print("found the destination node: ", node)
                break

            payload = self.adjacency_list[node["val"]]
            path_cost = 0

            if node or distances[node["val"]] != float("inf"):
                #  iterate over the list of edges for the current vertex
                for _, edge in enumerate(payload["edges"]):
                    # calculate the path cost at this point in the graph
                    # path cost = the value in distances table for previous node + the value of the edge between the previous node and current node
                    path_cost = distances[node["val"]] + edge["weight"]
                    #  if a shorter path has been found from the start vertex to the current vertex
                    if path_cost < distances[edge["node"]]:
                        # mark the new shorter distance to the current vertex in the distances dictionary
                        distances[edge["node"]] = path_cost
                        # mark the vertex that came before the current vertex
                        previous[edge["node"]] = node["val"]
                        # enqueue the current vertex with the updated path cost/ priority
                        queue.enqueue(edge["node"], path_cost)

    ########################                 for printing                 ##########################

    # time complexity: O(n)
    def printer(self):
        if self.adjacency_list:
            for key, value in self.adjacency_list.items():
                print(f"node: {key}  edges: {value}")


dijkstra_graph = Weighted_Graph()


dijkstra_graph.add_vertex("A")
dijkstra_graph.add_vertex("B")
dijkstra_graph.add_vertex("C")
dijkstra_graph.add_vertex("D")
dijkstra_graph.add_vertex("E")
dijkstra_graph.add_vertex("F")


dijkstra_graph.create_edge("A", "B", 4)
dijkstra_graph.create_edge("A", "C", 2)
dijkstra_graph.create_edge("B", "E", 3)
dijkstra_graph.create_edge("C", "D", 2)
dijkstra_graph.create_edge("C", "F", 4)
dijkstra_graph.create_edge("D", "E", 3)
dijkstra_graph.create_edge("D", "F", 1)
dijkstra_graph.create_edge("E", "F", 1)

# dijkstra_graph.remove_edge("D", "E")
# dijkstra_graph.remove_edge("D", "F")
# dijkstra_graph.remove_edge("C", "A")

# dijkstra_graph.remove_vertex("A")
# dijkstra_graph.remove_vertex("D")

print("finished: ", dijkstra_graph.shortest_path("A", "E"))

# dijkstra_graph.printer()
